### DISCUSSION TITLE
Crop color range

### DISCUSSION THREAD

**Discussion by Special-Collections** [ImageMagick/ImageMagick#6620] [Digital Image Processing]
Using ImageMagick 7.1.1.15 on Windows
Scripts do not run even with Cygwin installed

We have well over 1000 scans of illuminated medieval manuscripts we need to crop.
We crop away the yellowish paper background from:
https://freeimage.host/i/J970jA7
https://freeimage.host/i/J970wN9

To keep the colorful illuminated part of the images:
https://freeimage.host/i/J970Nte
https://freeimage.host/i/J970h9S

Note that the color and tone of the yellowish paper background may vary and
the position of the colorful illuminated part of the images may also vary.
Does anybody have an idea how to automate this process using ImageMagick?

**Comment by fmw42** [comment]
Use -trim to automatically remove outer areas of nearly the same color

`magick your_image.suffix -fuzz XX% -trim +repage result.suffix`

Adjust XX% as desired for color variation.  0 would be for perfectly pure color. Typical is 15%. Use more if your background is more varied.

**Reply by Special-Collections** [reply]
Thanks for your reply.
Does it imply we first need to draw a border of approximately the same color as the background we need to remove?
If so, how to draw such a border? The tone of various backgrounds may slightly differ.

**Reply by fmw42** [reply]
No.  However, your first image has a small black spot near the bottom left corner. So it prevents the -trim from getting past it.

`magick image.jpg -fuzz 25% -trim +repage result.jpg`

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/e5c22257-ead7-4f66-83bc-2971ea31a42a)


Your second image is cleaner, so this works

`magick image.jpg -fuzz 35% -trim +repage result.jpg`

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/27656516-1425-417a-aafa-37e4a019b142)

**Reply by fmw42** [reply]
For those image that have small dark spots, you can use -shave or -chop to remove that edge or all edges by some number of pixels to remove the spot and then -trim again.

`magick image.jpg -fuzz 35% -trim +repage -gravity south -chop 0x5 +repage -trim +repage result.jpg
`
![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/25658183-516c-4aa5-8123-0fce71603c2b)

**Comment by Special-Collections** [comment]
Thanks, perhaps anybody knows another way of doing this, a workaround?
-fuzz 20%% -fill white -floodfill +0+0 white  = also reaches the blank space between text lines that needs to stay intact
-background white -define trim:percent-background=0% -trim +repage = does not trim the darker areas of the background
We thought abount trimming a range of hex colors #C9CAC5 - #BAB690 from the edges.
Can ImageMagick find a range of hex colors?

**Reply by fmw42** [reply]
-trim does not need to know or have border color defined to run.  Flood fill with white does not trim, it only converts outer color to white if you have a close color that is the like the color at +0+0.  What more are you trying to do?

**Comment by Special-Collections** [comment]
Many of the scans have a "dirty" edge we need to remove and we also need to give the illuminated text a rectangular shape and move to the center op the page.
Before: https://freeimage.host/i/J9cE8bV
After: https://freeimage.host/i/J9cGjhN
Manually it would take a month of boring work to process 1400 scans.

**Reply by fmw42** [reply]
For such images, the only thing that comes to mind is AI/Deep Learning with lots of training.

**Reply by fmw42** [reply]
For this last image, I have a bash unix shell script for imagemagick that trims according to edge detail, if you are on Linux or Mac OSX, or Windows unix.

trim2detail` -a 100 image.jpg x.jpg`

![x](https://github.com/ImageMagick/ImageMagick/assets/13458218/a6094531-29e8-4ed4-b9a2-e0f3d51cb870)


see http://www.fmwconcepts.com/imagemagick

**Comment by snibgo** [comment]
> For this last image, I have a bash unix shell script ...

That script has trimmed detail from the top, while leaving the dirty edge on the right.

> Manually it would take a month of boring work to process 1400 scans.

I wonder how long it took to manually create those 1400 pages?

I don't see a simple automated solution for all the pages. The first image has faint decoration, and a more obtrusive black mark bottom-left. Some technique might identify the black mark as "not close" to other marks, and hence likely to be noise.

Obviously, don't start from JPEG images. I expect far better quality is available.

We can divide the problem into two sub-problems:

1. How much (if any) can we trim from the top and bottom?

2. How much (if any) can we trim from the left and right?

The following Windows BAT script demonstrates the approach. It does a difference of two blur to find where there is detail. Then it makes a vertical summary: scaling to one row, so each pixel is the mean of one column. Scale that back up to full size. "-grayscale Brightness" converts to gray, using the largest of the RGB values in each pixel. "auto-level" spreads the values to 0 to 100%, and we threshold at 4%. This give us v1.png, which has black columns where there is almost no detail and white where there is detail.

Similarly for a horizontal summary, written to h1.png.

For demonstration only, we make the black stripes semi-transparent and paste them over the input.
```
magick ^
  %INFILE% +write mpr:INP ^
  -set option:WW %%w ^
  -set option:HH %%h ^
  -blur 0x3 ^
  ( +clone -blur 0x3 ) ^
  -compose Difference -composite -auto-level ^
  ( -clone 0 -scale "%%[WW]x1^!" -scale "%%[WW]x%%[HH]^!" ^
    -grayscale Brightness ^
    -auto-level ^
    -threshold 4%% ^
+write v1.png ^
  ) ^
  ( -clone 0 -scale "1x%%[HH]^!" -scale "%%[WW]x%%[HH]^!" ^
    -grayscale Brightness ^
    -auto-level ^
    -threshold 4%% ^
+write h1.png ^
  ) ^
  -delete 0 ^
  -transparent White ^
  -channel A -evaluate Multiply 0.5 +channel ^
  mpr:INP ^
  -insert 0 ^
  -compose Over -layers Flatten ^
  %OUTFILE%
```

Here are the three results:

c1.jpg:
![c1](https://github.com/ImageMagick/ImageMagick/assets/33812368/5c580e64-b9b0-4efc-8709-a0453c55c2ae)

c2.jpg:
![c2](https://github.com/ImageMagick/ImageMagick/assets/33812368/167379a3-4991-4ff2-a29f-8c88bb6d5f9d)

c3.jpg:
![c3](https://github.com/ImageMagick/ImageMagick/assets/33812368/5312cf2b-8971-42b5-9472-2e44a1affc2d)

There are various possibilities for the next stage. For example, we might use the widest white stripes from v1.png and h1.png. I suggest you evaluate the command against a greater number of inputs.

Note that this command has not removed anything from the right side of the third image. As that is so noisy, it needs a different technique.

EDIT: 4% cuts off too much. Change both of these to 1% for a better result.

**Comment by snibgo** [comment]
For the noisy right-side of the third image: Is it always the right side? Is it ever on both sides? Does the width of the noise vary much?

For the three sample images, the following works. Crop to the right-hand 15% of the image. Convert to HCL, and consider just the C channel. This is Chroma, roughly "saturation". Find the average chroma in each column. Find the maximum of these values across the columns.

Windows BAT syntax:
```
magick ^
  %INFILE% +write mpr:INP ^
  -crop 100x100%%+%%[fx:w*0.85]+0 ^
  -set option:WW %%w ^
  -set option:HH %%h ^
  -colorspace HCL -channel 1 -separate +channel ^
  -format "offsetX = %%X\n" +write info: +format ^
  -scale "%%[WW]x1^!" ^
  -define identify:locate=maximum ^
  -define identify:limit=1 ^
  +write info: ^
  -define identify:locate=minimum ^
  -define identify:limit=1 ^
  info:
```

For image1, the maximum average chroma is 22.5%. For image2, it is 16.7%. For image3, it is 37.4%. So a threshold of 30% can be used. If the maximum average chroma is above 30%, we assume it has a noisy border that should be removed.

How much should we remove? We find the minimum average chroma. The position of this is where we should guillotine the right-hand third of the image. For the position within the original image, we need to add `offsetX`.

A script would need to extract data from the text outputs. For the three images, these are the text outputs:

image1:
```
offsetX = +722
Channel maximum locations:
  Gray: 13258.2 (0.202307) 13,0
Channel minimum locations:
  Gray: 9796.63 (0.149487) 76,0
```
The maximum is 0.202 (on a scale of 0.0 to 1.0), so it doesn't have a noisy border.

image2:
```
offsetX = +722
Channel maximum locations:
  Gray: 11675.3 (0.178154) 38,0
Channel minimum locations:
  Gray: 3806.17 (0.0580784) 0,0
```
The maximum is 0.178 (on a scale of 0.0 to 1.0), so it doesn't have a noisy border.

image3:
```
offsetX = +722
Channel maximum locations:
  Gray: 24511.4 (0.37402) 110,0
Channel minimum locations:
  Gray: 7744.27 (0.11817) 73,0
```
The maximum is 0.374 (on a scale of 0.0 to 1.0), so it does have a noisy border, which we need to remove.

We need to guillotine it at position 722 + 73 = 795.

```
%IMG7%magick manu3.jpg -crop 795x+0+0 +repage x.png
```
x.png is:
![x](https://github.com/ImageMagick/ImageMagick/assets/33812368/8bc4ece1-535d-4a93-9092-9c4a0bc5ec60)

Yes, that looks reasonable. But it should be tested on a number of inputs.

A similar process could be done on the left side.

**Reply by Special-Collections** [reply]
> minimum average chroma
You've found the clue, but the noise is not only on the right side and of various width.
Would be great if we could use the script in a Bat file on Windows

**Comment by Special-Collections** [comment]
First we were two, now we're three and we spent the summer vacation scanning medieval manuscripts.
Snigbo's script goes a long way and comes close to our desired result, but there is always a hitch.
We even tried to install Cygwin, but we can not run scripts on our available hardware.
If someone has time to run this script and a fast download connection this PDF contains 1212 scans.
Download link: https://we.tl/t-LxBgYXUaDq 175 MB
These scans were already -deskewed but unfortunately we do not have hi-res PNG images.
The other scans were already manually processed. 
This would surely save a few people several weeks of work.

**Comment by snibgo** [comment]
> but unfortunately we do not have hi-res PNG images.

Really? You carefully scan medieval manuscripts, and the images are JPEGs, most with just 502x709 pixels?

**Comment by Special-Collections** [comment]
Not always, in this particular case we just need the coloured decorative elements, the hi-res OCR text will later be pasted in.
PNG images would too much time to upload, download or otherwisely transfer.
It may be a great idea to find the northermost, southernmost, easternmost and westernmost positions that have minimum chroma and crop anything that lies beyond these positions. Do you have any ideas for a Windows Bat file?
Most saturated scan:
https://freeimage.host/i/J91WnKN
One of the least colourful scans:
https://freeimage.host/i/J91WolI
